tests/extended/test_trainer_ext.py

```diff
@@ -218,6 +230,43 @@ def test_run_seq2seq_slow(self):
         assert "generated_predictions.txt" in contents
         assert "predict_results.json" in contents
 
+    @slow
+    @require_bnb
+    def test_run_seq2seq_bnb_slow(self):
+        from transformers.training_args import OptimizerNames
+
+        def train_and_return_metrics(optim: str) -> Tuple[int, float]:
+            from pathlib import Path
+
+            extra_args = (
+                "--skip_memory_metrics 0 --optim {optim} --do_eval False --do_predict "
+                "False --adafactor False --log_level debug"
+            )
+
+            output_dir = self.run_trainer(
+                eval_steps=2,
+                max_len=128,
+                model_name=MARIAN_MODEL,
+                learning_rate=3e-4,
+                num_train_epochs=1,
+                distributed=False,
+                extra_args_str=extra_args.format(optim=optim),
+                do_eval=False,
+                do_predict=False,
+            )
+
+            # Check metrics
+            logs = TrainerState.load_from_json(Path(output_dir, "trainer_state.json")).log_history
+            gpu_peak_memory = logs[0]["train_mem_gpu_peaked_delta"]
+            loss = logs[0]["train_loss"]
+            return gpu_peak_memory, loss
+
+        original_gpu_peak_memory, original_loss = train_and_return_metrics(OptimizerNames.ADAMW_TORCH.value)
+        bnb_gpu_peak_memory, bnb_loss = train_and_return_metrics(OptimizerNames.ADAMW_BNB.value)
+
+        assert original_gpu_peak_memory < bnb_gpu_peak_memory
```

Comment:
do you have support for why checking that bnb peak memory is higher tests that bnb works?

I extended the code to dump the allocated delta as well, 

```
            gpu_peak_memory = logs[0]["train_mem_gpu_peaked_delta"]
            gpu_alloc_memory = logs[0]["train_mem_gpu_alloc_delta"]
            loss = logs[0]["train_loss"]
            return gpu_peak_memory, gpu_alloc_memory, loss

        original_gpu_peak_memory, original_gpu_alloc_memory, original_loss = train_and_return_metrics(OptimizerNames.ADAMW_TORCH.value)
        bnb_gpu_peak_memory, bnb_gpu_alloc_memory, bnb_loss = train_and_return_metrics(OptimizerNames.ADAMW_BNB.value)

        print(f"{original_gpu_alloc_memory=:012d} {original_gpu_peak_memory=:012d} {original_gpu_alloc_memory+original_gpu_peak_memory=:012d}")
        print(f"     {bnb_gpu_alloc_memory=:012d}      {bnb_gpu_peak_memory=:012d}           {bnb_gpu_alloc_memory+bnb_gpu_peak_memory=:012d}")
```

and I get:

```
original_gpu_alloc_memory=000645350400 original_gpu_peak_memory=000243983360 original_gpu_alloc_memory+original_gpu_peak_memory=000889333760
     bnb_gpu_alloc_memory=000507030528      bnb_gpu_peak_memory=000382303232           bnb_gpu_alloc_memory+bnb_gpu_peak_memory=000889333760
```

Looking at that total of 2 deltas which are identical tells me that both optimizers required the exact same amount of gpu memory, so it didn't save anything. or in other words a card with less memory wouldn't have been enough to run this training.

(you need py3.8 to have the `{foo=}` trick to work)

Reply:
or may be it's easier to look at the user-formatted stats:
```
AdamW:

***** train metrics *****
  before_init_mem_gpu        =        0MB
  init_mem_gpu_alloc_delta   =      206MB
  init_mem_gpu_peaked_delta  =        0MB
  train_mem_gpu_alloc_delta  =      615MB
  train_mem_gpu_peaked_delta =      232MB
  
  BNB:
  
  
***** train metrics *****
  before_init_mem_gpu        =        0MB
  init_mem_gpu_alloc_delta   =      206MB
  init_mem_gpu_peaked_delta  =        0MB
  train_mem_gpu_alloc_delta  =      483MB
  train_mem_gpu_peaked_delta =      364MB
  
```

Reply:
Here is how I understand the work of BNB optimizer. Let's look at a single train iteration and how much memory optimizer states consume:

1. `forward`: 2 bytes
2. `backward`: 2 bytes
3. `step`: dequantized to 8 bytes

Assuming that `forward` activation memory has been released when `backward` has finished there should be some "spare" peak memory available to expand 2 to 8 bytes per paper to perform the step. I think in general it should be less than the additional 6 bytes per param. So the optimizer step is very likely to consume more "peak" memory.

But the test you wrote shows that we end up with the exact same amount of total gpu memory used by the card to perform a training step including peak memory. Which defeats the purpose of BNB.

Am I missing something in my understanding of the algorithm?

Reply:
@TimDettmers, am I missing something here? I have rewatched your video and skimmed through the paper but somehow I'm missing the discussion of 6 additional bytes materialization during step. Thank you!

Reply:
Everything that you write is correct in terms of how 8-bit optimizers work. There can be several issues why you do not see a decrease in memory.

1. PyTorch _does not release memory_ to enable more efficient future allocation by reusing old tensor chunks for allocation. This is why if you track nvidia-smi or allocated memory statistics you will not see a difference. To track the difference you need to look at used tensors (not allocated ones). PyTorch has a function for this.
2. If the weight tensor is below a certain size (I think it was 8192), 32-bit optimizers are used by default and no memory is saved.

Reply:
> Everything that you write is correct in terms of how 8-bit optimizers work. 

Could you please confirm whether it staggers the dequantization and quantization so that only one param is processed at the time, and thus requiring `6x largest param numel` of additional gpu memory bytes, or does it do it for all params in one go and thus require `6x model params`?

> There can be several issues why you do not see a decrease in memory.
>  1. PyTorch _does not release memory_ to enable more efficient future allocation by reusing old tensor chunks for allocation. This is why if you track nvidia-smi or allocated memory statistics you will not see a difference. To track the difference you need to look at used tensors (not allocated ones). PyTorch has a function for this.

The memory tracker isn't using `nvidia-smi` (or nvml) but torch memory management API including clearing cache, see:
https://github.com/huggingface/transformers/blob/7c5d79912a21880ce13d77881940458e90d98917/src/transformers/trainer_utils.py#L331
what other pytorch function are you referring to?

I will probably need to try to pre-load the gpu memory so that only enough remains for it to run `-6 bytes` per param and see if this forces pytorch to reuse the cached memory.
 
>     2. If the weight tensor is below a certain size (I think it was 8192), 32-bit optimizers are used by default and no memory is saved.

Since the allocated and peak memory pairs are very different with and without bnb, I don't think we are hitting this particular case.

Any suggestions at how to write such a test? 

I'm thinking perhaps forcing to run it on `cpu` and measuring peak RSS, but usually cpu memory is even more difficult to measure correctly than gpu memory due to linux's super complicated CPU memory management.

Reply:
> Could you please confirm whether it staggers the dequantization and quantization cycle during `step` so that only one param is processed at the time, and thus requiring `6x largest param numel` of additional gpu memory bytes, or does it do it for all params in one go and thus require `6x total params numel`?

Oh, this actually works differently. Expansion to 32-bit is done in cache and never on the GPU memory level, as such only the 8-bit tensor exist at the PyTorch level and never any 16/32-bit copy.

> The memory tracker isn't using `nvidia-smi` (or nvml) but torch memory management API including clearing cache, see:
> 
> https://github.com/huggingface/transformers/blob/7c5d79912a21880ce13d77881940458e90d98917/src/transformers/trainer_utils.py#L331

My guess is that some buffers are freed but their memory is not released and then these buffers are reused by the 8-bit optimizer buffers. But even with this there should be some memory decrease.

> Except of course it doesn't know to run `torch.cuda.empty_cache()` before `step`. I guess I could manually insert it just for the sake of seeing that it works.

Unlike standard Adam optimziers, the 8-bit optimizers memory for the buffers is only allocated during the first `step()` call. Might this cause the problem? I this is so, there might be some success to call `torch.cuda.empty_cache()` before the `step()` call. It seems that `empty_cache()` does not release memory chunks which have already been reused, so calling it some time before `step()` might be the right way to do it.

> What other pytorch function are you referring to? I'm using `torch.cuda.memory_allocated` and not `torch.cuda.memory_reserved`: https://pytorch.org/docs/stable/notes/cuda.html#memory-management

I do not know the details. It seems `max_memory_allocated()` + `empty_cache()` is the right way to do it. `memory_reserved()` might be able to help to isolated which tensors are reused (for the 8-bit optimizer states) and which are cached.

> Any suggestions at how to write such a test?

I think if one can track reserved + allocated memory before the first `step()` and after it, one probably can determine what memory is allocated. Probably it is easiest and best, to make use of `empty_cache()`, but as I wrote above, this does not free memory blocks which have been reallocated and these might be partially reused for 8-bit states which might not lower the memory footprint. I think its best to use `empty_cache()` whenever some new dynamic memory is allocated (other than model parameters), so to keep the reserved memory to a minimum, then the memory benefits of 8-bit optimizers should be visible I think. So probably, call `empty_cache()` after `forward()` after `backward()` and then before `step()` in the case there are some other things going on in the background. I think with that it should work.

Reply:
Thank you for the detailed answers, Tim.

Obviously we can't / shouldn't do any of the `empty_cache()` calls for the normal functionality, but we are discussing using these just to prove that it works. The current test makes a call to an external HF Trainer application which can't be modified, but perhaps we could use a custom small training loop in a test and measure the gpu memory there, then we will have a full control over gpu memory.

Once it becomes clear which bits accounting indicates the 8-bit optimizer use then we should be able to write a test.

> > Could you please confirm whether it staggers the dequantization and quantization cycle during `step` so that only one param is processed at the time, and thus requiring `6x largest param numel` of additional gpu memory bytes, or does it do it for all params in one go and thus require `6x total params numel`?
> 
> Oh, this actually works differently. Expansion to 32-bit is done in cache and never on the GPU memory level, as such only the 8-bit tensor exist at the PyTorch level and never any 16/32-bit copy.

Sorry, I'm still not having full clarity here - which "cache" are you referring here? something in CPU memory?

Or perhaps a link to a source is worth thousands words?

Reply:
This is now very clear, Tim! Thanks a lot!

Let me try and see if I can write a telling test.

Reply:
OK, so actually the problem wasn't in the test itself. It was because the test was not using a fresh process - and when this is happening the outcome is completely broken. I switched it to using an external process and now it shows the memory difference in a very loud and clear way.

I extended the test to do multiple checks including an absolute 6x saving.

(Also had to fix the rebasing as files got moved around, but I think I got it all right)

cc: @manuelciosici